packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs

```diff
@@ -641,6 +617,91 @@ where
             .attach_printable_lazy(|| uri.clone())?
             .get(0))
     }
+
+    /// Inserts a [`Link`] associated with an [`AccountId`], into the database.
+    ///
+    /// # Errors
+    ///
+    /// - if the [`Link`] exists already
+    /// - if the [`Link`]s link type doesn't exist
+    /// - if inserting the link failed.
+    async fn insert_link(&self, link: &Link, created_by: AccountId) -> Result<(), InsertionError> {
+        let link_type_version_id = self
+            .version_id_by_uri(link.link_type_uri())
+            .await
+            .change_context(InsertionError)
+            .attach_printable(link.source_entity())?;
+
+        self.as_client()
+        .query_one(
+            r#"
+            INSERT INTO links (source_entity_id, target_entity_id, link_type_version_id, link_order, created_by)
+            VALUES ($1, $2, $3, null, $4)
+            RETURNING source_entity_id, target_entity_id, link_type_version_id;
+            "#,
+            &[&link.source_entity(), &link.target_entity(), &link_type_version_id, &created_by],
+        )
+        .await
+        .into_report()
+        .change_context(InsertionError)
+        .attach_printable(created_by)
+        .attach_lazy(|| link.clone())?;
+
+        Ok(())
+    }
+
+    /// Moves a [`Link`] associated with an [`AccountId`] from the `links` table into the
+    /// `link_histories` table.
+    ///
+    /// # Errors
+    ///
+    /// - if the [`Link`] doesn't exist
+    /// - if the [`Link`]s link type doesn't exist
+    /// - if inserting the link failed.
+    async fn move_link_to_history(
+        &self,
+        link: &Link,
+        removed_by: AccountId,
+    ) -> Result<(), LinkRemovalError> {
+        let link_type_version_id = self
+            .version_id_by_uri(link.link_type_uri())
+            .await
+            .change_context(InsertionError)
+            .attach_printable(link.source_entity())
+            .change_context(LinkRemovalError)?;
+
+        self.as_client()
+            .query_one(
+                // This query removes a link from the `links` table and then immediately inserts
+                // into the link_histories table.
+                r#"
+                WITH removed AS (
+                    DELETE FROM links
```

Comment:
Hm when we discussed this I don't think I'd thought it through fully that we'd actually be deleting records in the DB in the interim.

Perhaps this is fine. Would be good to get @benwerner01's thoughts when he's back

Reply:
Yep, it does cross that line of actually deleting. I would like to think that since this happens in one step within a transaction (and `query_one`) we expect an insertion for any deletion

Reply:
As I understand from @TimDiekmann part of the reasoning for introducing a new table (rather than using a `deleted_at` column to determine whether a link has been deleted) is because we would not be able to preserve the history of a link that is added, removed and re-added. This makes sense and is definitely fine for now, but I think we may need to have a discussion around why we have chosen not to associate identifiers with links in the datastore, which would've have prevented this issue.